src/test/crimson/test_socket.cc

```diff
@@ -0,0 +1,203 @@
+};
+
+class AcceptTest final
+    : public SocketFactoryBase<AcceptTest> {
+ public:
+  future<> handle_server_socket(SocketFRef&& socket) override {
+    return seastar::sleep(100ms
+    ).then([socket = std::move(socket)] () mutable {
+      return socket->close()
+      .finally([socket = std::move(socket)] {});
+    });
+  }
+};
+
+future<> test_refused() {
+  logger().info("test_refused()...");
+  return AcceptTest::connect().discard_result(
+  ).then([] {
+    ceph_abort_msg("connection is not refused");
+  }).handle_exception_type([] (const std::system_error& e) {
+    if (e.code() != error::connection_refused) {
+      logger().error("test_refused() got unexpeted error {}", e);
+      ceph_abort();
+    }
+    // successful
+  }).handle_exception([] (auto eptr) {
+    logger().error("test_refused() got unexpeted exception {}", eptr);
+    ceph_abort();
+  });
+}
+
+future<> test_bind_same() {
+  logger().info("test_bind_same()...");
+  return ceph::net::create_sharded<AcceptTest>(
+  ).then([] (AcceptTest* factory) {
+    return factory->bind_accept(
+    ).then([] {
+      // try to bind the same address
+      return ceph::net::create_sharded<AcceptTest>(
+      ).then([] (AcceptTest* factory2) {
+        return factory2->bind_accept(
+        ).then([] {
+          ceph_abort_msg("bind should raise addr-in-use");
+        }).finally([factory2] {
+          return factory2->shutdown();
+        }).handle_exception_type([] (const std::system_error& e) {
+          if (e.code() != error::address_in_use) {
+            logger().error("test_bind_same() got unexpeted error {}", e);
+            ceph_abort();
+          }
+          // successful
+        });
+      });
+    }).finally([factory] {
+      return factory->shutdown();
+    }).handle_exception([] (auto eptr) {
+      logger().error("test_bind_same() got unexpeted exception {}", eptr);
+      ceph_abort();
+    });
+  });
+}
+
+future<> test_accept() {
+  logger().info("test_accept()");
+  return ceph::net::create_sharded<AcceptTest>(
+  ).then([] (AcceptTest* factory) {
+    return factory->bind_accept().then([factory] {
+      return seastar::when_all(
+        factory->connect().then([] (auto socket) {
+          return socket->close().finally([cleanup = std::move(socket)] {}); }),
+        factory->connect().then([] (auto socket) {
+          return socket->close().finally([cleanup = std::move(socket)] {}); }),
+        factory->connect().then([] (auto socket) {
+          return socket->close().finally([cleanup = std::move(socket)] {}); })
+      ).discard_result();
+    }).then([] {
+      // should be enough to be connected locally
+      return seastar::sleep(50ms);
+    }).finally([factory] {
+      return factory->shutdown();
+    }).handle_exception([] (auto eptr) {
+      logger().error("test_accept() got unexpeted exception {}", eptr);
+      ceph_abort();
+    });
+  });
+}
+
+}
+
+int main(int argc, char** argv)
```

Comment:
Seastar offers (very) simple infrastructure for unit-testing under `seastar::testing`. I haven't tried it out, just signalizing its presence for the future.

The socket test LGTM :+1: